tests/test_grammar.py

```diff
@@ -137,6 +137,30 @@
     ("str_quoted_too_many_1", "''a'", GrammarParseError),
     ("str_quoted_too_many_2", "'a''", GrammarParseError),
     ("str_quoted_too_many_3", "''a''", GrammarParseError),
+    ("str_quoted_trailing_esc_1", "'abc\\\\'", "abc\\"),
+    ("str_quoted_trailing_esc_2", "'abc\\\\\\\\'", "abc\\\\"),
+    ("str_quoted_no_esc_1", r'"abc\def"', r"abc\def"),
+    ("str_quoted_no_esc_2", r'"abc\\def"', r"abc\def"),
+    ("str_quoted_no_esc_3", r'"\\\abc\def"', r"\\abc\def"),
+    ("str_quoted_bad_1", r'"abc\"', GrammarParseError),
+    ("str_quoted_bad_2", r'"abc\\\"', GrammarParseError),
+    ("str_quoted_esc_quote_single_1", r"'abc\'def'", "abc'def"),
+    ("str_quoted_esc_quote_single_2", r"'abc\\\'def'", r"abc\'def"),
+    ("str_quoted_esc_quote_single_3", r"'abc\\\\\'def'", r"abc\\'def"),
+    ("str_quoted_esc_quote_single_4", r"'a\'b\'cdef\\\''", r"a'b'cdef\'"),
+    ("str_quoted_esc_quote_single_bad", r"'abc\\'def'", GrammarParseError),
+    ("str_quoted_esc_quote_double_1", r'"abc\"def"', 'abc"def'),
+    ("str_quoted_esc_quote_double_2", r'"abc\\\"def"', r"abc\"def"),
+    ("str_quoted_esc_quote_double_3", r'"abc\\\\\"def"', r'abc\\"def'),
+    ("str_quoted_esc_quote_double_4", r'"a\"b\"cdef\\\""', r'a"b"cdef\"'),
+    ("str_quoted_esc_quote_double_bad", r'"abc\\"def"', GrammarParseError),
+    ("str_quoted_empty", "''", ""),
+    ("str_quoted_basic", "'a'", "a"),
+    ("str_quoted_tmp_1", r"'\a'", r"\a"),
+    ("str_quoted_tmp_2", r"'a\'", GrammarParseError),
+    ("str_quoted_inside_quote_different", "'\"'", '"'),
+    ("str_quoted_inside_quote_same", r"'\''", "'"),
+    ("str_quoted_extra_quote", r"'c:\\''", GrammarParseError),
```

Comment:
This works?
```suggestion
    ("str_quoted_trailing_backslash", r"'c:\\'", "c:\"),
    ("str_quoted_extra_quote", r"'c:\\''", GrammarParseError),
```

Loos like rstrings are working for it:
```
print(r"'c:\\'")
'c:\\'
```
I would expect `\\` to be interpreted as a token equal to `\`, so the content of the quoted string should be `c:\`

Reply:
That would work (after fixing `"c:\"` which isn't a valid Python string) but there's already a similar test so no need to add it:
```python
("str_quoted_trailing_esc_1", "'abc\\\\'", "abc\\"),
```

r-strings work as long as the "escaped" quote isn't the same type as the enclosing quote.

Reply:
That test is similar, but is testing for a double trailing backslash.
Since this is all about backslash and escaping, I think an additional basic test for a single backslash is worth it. 

And now I realize that the double backslash is due to Python interpretation of escaping.
Why not use r-strings here?
```python
In [3]: print(r"'abc\\'")
'abc\\'
```

Reply:
Yeah it's a bit tricky to get those `\` right :)

> Why not use r-strings here?

I explained my reasoning in https://github.com/omry/omegaconf/pull/630#discussion_r613615877 (I don't mind changing it though if you like the "partial" r-string version better)

Reply:
yeah, let's use r-strings whenever possible, where it's not possible add a comment.
This part of the tests could be better tested by reading raw data from a file instead of encoding the data data in Python but I am not sure it's worth it to do it just for the escaping.

Reply:
Ok, done in 9011928
I used a couple of tricks to use raw strings everywhere -- I can revert those if you don't like them:

Using `.strip()` for strings ending in with a `\`, e.g.
```python
("str_quoted_trailing_esc_1", r"'abc\\'", r" abc\ ".strip()),
```
(technically the leading whitespace is not needed, but I find it looks better)

Using an f-string to include tabs in r-strings, e.g.
```python
("str_esc_ws_2", fr"\ \{TAB}\{TAB}", f" {TAB}{TAB}"),
```
where `TAB` is a constant defined as
```python
TAB = "\t"  # to be used in raw strings, e.g. `fr"C:\{TAB}foo"`
```

Reply:
When reviewing I was scratching my head a bit before figuring out the strip trick.
I think it's fine, but add a comment.
maybe something like:
```
# raw strings does not allow trailing \, adding a space and stripping it.
```

Reply:
Done in 6015a71